File: server/src/main/java/org/hkurh/doky/documents/DocumentController.java
package org.hkurh.doky.documents;


import org.hkurh.doky.security.DokyAuthority;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.annotation.Secured;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.io.IOException;
import java.net.MalformedURLException;

import static java.lang.String.format;

@RestController
@RequestMapping("/documents")
@Secured(DokyAuthority.Role.ROLE_USER)
public class DocumentController implements DocumentApi {
    private static final Logger LOG = LoggerFactory.getLogger(DocumentController.class);

    private DocumentFacade documentFacade;

    @Override
    @PostMapping("/{id}/upload")
    public ResponseEntity<?> uploadFile(@RequestBody MultipartFile file, @PathVariable String id) {
        getDocumentFacade().saveFile(file, id);
        return ResponseEntity.ok(null);
    }

    @Override
    @GetMapping("/{id}/download")
    public ResponseEntity<?> downloadFile(@PathVariable String id) throws IOException {
        try {
            var file = getDocumentFacade().getFile(id);
            if (file == null) {
                LOG.debug(format("No attached file for document [%s]", id));
                return ResponseEntity.noContent().build();
            }
            var header = "attachment; filename=\"" +
                    file.getFilename() +
                    "\"";
            var response = ResponseEntity.ok()
                    .contentType(MediaType.APPLICATION_OCTET_STREAM)
                    .header(HttpHeaders.CONTENT_DISPOSITION, header)
                    .body(file);
            return response;
        } catch (MalformedURLException e) {
            return ResponseEntity.internalServerError().build();
        }
    }

    @Override
    @PostMapping
    public ResponseEntity<?> create(@RequestBody DocumentRequest document) {
        var createdDocument = getDocumentFacade().createDocument(document.getName(), document.getDescription());
        var resourceLocation = ServletUriComponentsBuilder.fromCurrentRequest()
                .path("/{id}").build(createdDocument.getId());

        return ResponseEntity.created(resourceLocation).build();
    }

    @Override
    @GetMapping
    public ResponseEntity<?> getAll() {
        var documents = getDocumentFacade().findAllDocuments();
        if (documents.isEmpty()) {
            LOG.debug("No Documents for current user");
            return ResponseEntity.noContent().build();
        } else {
            return ResponseEntity.ok(documents);
        }
    }

    @Override
    @GetMapping("/{id}")
    public ResponseEntity<?> get(@PathVariable String id) {
        var document = getDocumentFacade().findDocument(id);
        if (document != null) {
            return ResponseEntity.ok(document);
        } else {
            LOG.debug(format("No Document with id [%s]", id));
            return ResponseEntity.noContent().build();
        }
    }

    public DocumentFacade getDocumentFacade() {
        return documentFacade;
    }

    @Autowired
    public void setDocumentFacade(DocumentFacade documentFacade) {
        this.documentFacade = documentFacade;
    }
}

File: server/src/main/java/org/hkurh/doky/documents/DocumentController.kt
package org.hkurh.doky.documents

import org.hkurh.doky.security.DokyAuthority
import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.access.annotation.Secured
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import java.io.IOException
import java.net.MalformedURLException

@RestController
@RequestMapping("/documents")
@Secured(DokyAuthority.Role.ROLE_USER)
class DocumentController(private val documentFacade: DocumentFacade) : DocumentApi {
    @PostMapping("/{id}/upload")
    override fun uploadFile(@RequestBody file: MultipartFile?, @PathVariable id: String?): ResponseEntity<*> {
        documentFacade.saveFile(file, id)
        return ResponseEntity.ok<Any>(null)
    }

    @GetMapping("/{id}/download")
    @Throws(IOException::class)
    override fun downloadFile(@PathVariable id: String?): ResponseEntity<*> {
        return try {
            val file = documentFacade.getFile(id)
            if (file == null) {
                LOG.debug(String.format("No attached file for document [%s]", id))
                return ResponseEntity.noContent().build<Any>()
            }
            val header = "attachment; filename=\"" +
                    file.filename +
                    "\""
            ResponseEntity.ok()
                    .contentType(MediaType.APPLICATION_OCTET_STREAM)
                    .header(HttpHeaders.CONTENT_DISPOSITION, header)
                    .body(file)
        } catch (e: MalformedURLException) {
            ResponseEntity.internalServerError().build<Any>()
        }
    }

    @PostMapping
    override fun create(@RequestBody document: DocumentRequest?): ResponseEntity<*> {
        val createdDocument = documentFacade.createDocument(document!!.name, document.description)
        val resourceLocation = ServletUriComponentsBuilder.fromCurrentRequest()
                .path("/{id}").build(createdDocument!!.id)
        return ResponseEntity.created(resourceLocation).build<Any>()
    }

    @get:GetMapping
    override val all: ResponseEntity<*>
        get() {
            val documents = documentFacade.findAllDocuments()
            return if (documents!!.isEmpty()) {
                LOG.debug("No Documents for current user")
                ResponseEntity.noContent().build<Any>()
            } else {
                ResponseEntity.ok(documents)
            }
        }

    @GetMapping("/{id}")
    override fun get(@PathVariable id: String?): ResponseEntity<*> {
        val document = documentFacade.findDocument(id)
        return if (document != null) {
            ResponseEntity.ok(document)
        } else {
            LOG.debug(String.format("No Document with id [%s]", id))
            ResponseEntity.noContent().build<Any>()
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(DocumentController::class.java)
    }
}
